### PR TITLE
Fix for issue #603 - Error 500 when creating a db below quorum

### PR DESCRIPTION
**Pull request by jjrodrig**
## Overview
Current behaviour of DB creation in a cluster degraded situation is not consistent with the general behaviour described for document creation in the same situation.

> The number of copies of a document with the same revision that have to be read before CouchDB returns with a 200 is equal to a half of total copies of the document plus one. It is the same for the number of nodes that need to save a document before a write is returned with 201. If there are less nodes than that number, then 202 is returned. Both read and write numbers can be specified with a request as r and w parameters accordingly.

The current behaviour for database creation in a cluster is:
 - Database creation returns 201 - Created if all nodes responds ok
 - Database creation returns 202 - Accepted if the quorum is met
 - Database creation returns 500 - Error if the responses are bellow quorum

The quorum is the default: Number of nodes/2 +1

This PR changes the database creation result with the following behaviour:
- Database creation returns 201 - Creation if the quorum is met
- Database creation returns 202 - Accepted if at least one node responds ok
- Database creation returns 500 - Error if there is no correct response from any node

## Testing recommendations

- All previous tests are ok
- I've focused on chttpd and javascript tests 
- I've skiped reduce_builtin.js test as it is failing even in master branch
`test/javascript/tests/reduce_builtin.js                        
    Error: {gen_server,call,[<0.2426.1>,{get_state,49},infinity]}`

`make check apps=chttpd ignore_js_suites=reduce_builtin`

I didn't identify testing infrastructure for testing cluster degradation issues. So I decided to add some support for testing in different cluster conditions. Now make has two more task:
- test-cluster-with-quorum, which launch a three node cluster and stops one node then executes the tests
- test-cluster-without-quorum which launch a three node cluster and stops two nodes then executes the tests

This PR can be tested in this way
```
 make test-cluster-with-quorum
 make test-cluster-without-quorum
```

## Related Issues or Pull Requests
Issue #603 

## Checklist

- [x] Code is written and works correctly;
- [x] Changes are covered by tests; (DB creation yes, Cluster degradation no)
- [ ] Documentation reflects the changes;

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -129,6 +129,40 @@ endif
...
+	@mkdir -p share/www/script/test
+ifeq ($(IN_RELEASE), true)
+	@cp test/javascript/tests/lorem*.txt share/www/script/test/
+else
+	@mkdir -p src/fauxton/dist/release/test
+	@cp test/javascript/tests/lorem*.txt src/fauxton/dist/release/test/
+endif
+	@rm -rf dev/lib
+	@dev/run -n 3 -q --with-admin-party-please \
+						--enable-erlang-views --degrade-cluster 2 \
```
Comment (rnewson) [inline Makefile]: indentation

File: Makefile [hunk 2]
```diff
@@ -129,6 +129,40 @@ endif
...
+	@mkdir -p src/fauxton/dist/release/test
+	@cp test/javascript/tests/lorem*.txt src/fauxton/dist/release/test/
+endif
+	@rm -rf dev/lib
+	@dev/run -n 3 -q --with-admin-party-please \
+            --enable-erlang-views --degrade-cluster 1 \
+            -c 'startup_jitter=0' \
+            'test/javascript/run --suites "$(suites)" \
+            --ignore "$(ignore_js_suites)" \
+						--path test/javascript/tests-cluster/with-quorum'
```
Comment (rnewson) [inline Makefile]: indentation

File: src/fabric/src/fabric_db_create.erl [hunk 3]
```diff
@@ -146,9 +146,9 @@ maybe_stop(W, Counters) ->
         {ok, {W, Counters}};
     false ->
         case lists:sum([1 || {_, ok} <- Counters]) of
-        W ->
+        NumOk when NumOk >= (W div 2 +1) ->
             {stop, ok};
-        NumOk when NumOk >= (W div 2 + 1) ->
+        NumOk when NumOk > 0 ->
```
Comment (rnewson) [inline src/fabric/src/fabric_db_create.erl]: this is the right change.

**Comment by rnewson** [comment]
+1

**Comment by janl** [comment]
@jjrodrig we wanted to make it so that `make test-cluster-with[out]-quorum` is run as part of `make check`, but this doesn’t seem to work on Travis, would you be able to assist with debugging this?

**Comment by jjrodrig** [comment]
Ok, I'll check it.

**Comment by janl** [comment]
Thank you! :)

**Comment by janl** [comment]
I’m trying a new variant where the new make targets run first in `make check`

**Comment by jjrodrig** [comment]
@janl I've reproduced the problem in my environment. It seems that the stop after the mango test is collisioning with the start of the cluster test. 
If you see this commit:

https://github.com/jjrodrig/couchdb/commit/f45b404de930b7da81af7e01d2071451d21acde7

I've introduced an sleep after mango tests and the check is working

Also, I've noticed that the mango test do not clean up dev/lib so it can be affected by previous test executions.

**Comment by janl** [comment]
merged this, and filed an issue for cleaning up mango: https://github.com/apache/couchdb/issues/1134

Thanks a lot for this contribution!

**Pull request merged**